src/variable_info_private.hpp

```diff
@@ -0,0 +1,487 @@
+	void do_from_config(
+		typename maybe_const<config, V>::reference cfg, typename get_variable_key_visitor::param_t state) const
+	{
+		state.type_ = state_named;
+		state.key_ = key_;
+		state.child_ = &cfg;
+	}
+
+	const std::string& key_;
+};
+
+/**
+ * Appends an [index] to the variable.
+ * We only support from_named since [index][index2] or a.length[index] both don't make sense.
+ */
+template<typename V>
+class get_variable_index_visitor : public info_visitor<V, void>
+{
+public:
+	get_variable_index_visitor(int n)
+		: n_(n)
+	{
+	}
+
+	virtual void from_named(typename get_variable_index_visitor::param_t state) const override
+	{
+		state.index_ = n_;
+		resolve_negative_value(state.child_->child_count(state.key_), state.index_);
+		state.type_ = state_indexed;
+	}
+
+private:
+	const int n_;
+};
+
+/** Tries to convert it to an (maybe const) attribute value. */
+template<typename V>
+class as_scalar_visitor : public info_visitor_const<V, typename maybe_const<config::attribute_value, V>::reference>
+{
+public:
+	virtual typename as_scalar_visitor::result_t from_named(
+            typename as_scalar_visitor::param_t state) const override
+	{
+		return (*state.child_)[state.key_];
+	}
+
+	/**
+	 * Only implemented for read-only variable_info. Other types use the default throw implementation.
+	 */
+	virtual typename as_scalar_visitor::result_t from_temporary(
+	        typename as_scalar_visitor::param_t /*state*/) const override
+	{
+		throw invalid_variablename_exception();
+	}
+};
+
+/**
+ * Values like '.length' are readonly so we only support reading them, especially since we don't
+ * want to return non-const references.
+ */
+template<>
+const config::attribute_value& as_scalar_visitor<const vi_policy_const>::from_temporary(
+	as_scalar_visitor::param_t state) const
+{
+	return state.temp_val_;
+}
+
+/**
+ * Tries to convert to a [const] config&. Unlike range based operation this also supports 'from_start'.
+ * NOTE: Currently getting the 'from_start' case here is impossible, because we always apply at least one string key.
+ */
+template<typename V>
+class as_container_visitor : public info_visitor_const<V, typename maybe_const<config, V>::reference>
+{
+public:
+	virtual typename as_container_visitor::result_t from_named(
+	        typename as_container_visitor::param_t state) const override
+	{
+		return V::get_child_at(*state.child_, state.key_, 0);
+	}
+
+	virtual typename as_container_visitor::result_t from_start(
+	        typename as_container_visitor::param_t state) const override
+	{
+		return *state.child_;
+	}
+
+	virtual typename as_container_visitor::result_t from_indexed(
+	        typename as_container_visitor::param_t state) const override
+	{
+		return V::get_child_at(*state.child_, state.key_, state.index_);
+	}
+};
+
+/**
+ * This currently isn't implemented as a range-based operation because doing it on something like range
+ * 2-5 on vi_policy_const if child_ has only 4 elements would be too hard to implement.
```

Comment:
Okay, this is another reason why you shouldn't include formatting changes and functional changes in the same commit. It makes it impossible to tell what comes from your changes and what was there already.